File: frontend/src/useCases/sendMessage.ts
import Preferences from "../entities/Preferences/Preferences"

import { SendMessage } from '../../wailsjs/go/main/App'
import { ipc } from '../../wailsjs/go/models'
import { LogPrint } from "../../wailsjs/runtime/runtime"

type sendMessageRequest = {
  chatId: string,
  content: string,
}

const sendMessage = async (props: sendMessageRequest) => {
  const messageProps = {
    id: Math.random().toString(),
    chatId: props.chatId,
    content: props.content,
    senderUserId: Preferences.userId,
    datetime: new Date()
  }

  const { chatId, content, senderUserId, datetime } = messageProps
  const message = SendMessage(
    new ipc.SendMessageRequest({
      id: 'vvsvsfvfs',
      chatId,
      content,
      senderUserId,
      datetime: datetime.toString()
    })
  )

    LogPrint("messageREquest returned from GO:  " + JSON.stringify(await message, null, 2))
}

export default sendMessage

File: ipc/SendMessage.go
package ipc

import (
	app "tzat/core/App"
	chat "tzat/core/Chat"

	"github.com/google/uuid"
	"github.com/wailsapp/wails/v2/pkg/runtime"
)

type SendMessageRequest struct {
	ChatId       string `json:"chatId"`
	Content      string `json:"content"`
	Datetime     string `json:"datetime"`
	SenderUserId string `json:"senderUserId"`
}

func (c *Channel) SendMessage(messageRequest SendMessageRequest) SendMessageRequest {
	runtime.EventsEmit(app.GetInstance().Ctx, "receivedMessage", messageRequest)

	chatCollection := chat.GetCollection()
	newMessage := chat.Message{
		Id:           uuid.New().String(), // TODO: the id will be handled by the database once the server side is implemented
		ChatId:       messageRequest.ChatId,
		Content:      messageRequest.Content,
		Datetime:     messageRequest.Datetime,
		SenderUserId: messageRequest.SenderUserId,
	}
	chatCollection.AddMessageToChat(newMessage)

	return messageRequest
}
